tikv-client/src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java

```diff
@@ -127,163 +128,207 @@ public void resolve() {
+        colBuilder.setColumnLen(columnInfo.getColumnLen());
+        colBuilder.setFlag(columnInfo.getFlag());
+        if (columnInfo.getColumnId() == -1) {
+          hasPk = true;
+          colBuilder.setPkHandle(true);
+        }
+        indexScanBuilder.addColumns(colBuilder);
       }
-      indexScanBuilder.addColumns(colBuilder);
-    }
 
-    if (!hasPk) {
-      ColumnInfo handleColumn = ColumnInfo.newBuilder()
-          .setColumnId(-1)
-          .setPkHandle(true)
-          // We haven't changed the field name in protobuf file, but
-          // we need to set this to true in order to retrieve the handle,
-          // so the name 'setPkHandle' may sounds strange.
-          .build();
-      indexScanBuilder.addColumns(handleColumn);
-    }
-    executorBuilder.setTp(ExecType.TypeIndexScan);
-
-    indexScanBuilder
-        .setTableId(tableInfo.getId())
-        .setIndexId(indexInfo.getId());
-    dagRequestBuilder.addExecutors(executorBuilder.setIdxScan(indexScanBuilder).build());
-    int colCount = indexScanBuilder.getColumnsCount();
-    dagRequestBuilder.addOutputOffsets(
-         colCount != 0 ? colCount - 1 : 0
-    );
-    return dagRequestBuilder
-        .setFlags(flags)
-        .setTimeZoneOffset(timeZoneOffset)
-        .setStartTs(startTs)
-        .build();
-  }
+      if (isDoubleRead()) {
+        // double read case
+        if (!hasPk) {
+          indexScanBuilder.addColumns(handleColumn());
+        }
 
-  /**
-   * @return DAGRequest built
-   */
-  private DAGRequest buildTableScan() {
-    checkArgument(startTs != 0, "timestamp is 0");
-    DAGRequest.Builder dagRequestBuilder = DAGRequest.newBuilder();
-    Executor.Builder executorBuilder = Executor.newBuilder();
-    TableScan.Builder tblScanBuilder = TableScan.newBuilder();
+        int colCount = indexScanBuilder.getColumnsCount();
+        // double read case: need to retrieve handle
+        dagRequestBuilder.addOutputOffsets(
+            colCount != 0 ? colCount - 1 : 0
+        );
+      } else {
+        int count = 0;
+        int colCount = indexScanBuilder.getColumnsCount();
+        boolean pkIsNeeded = false;
+        // =================== IMPORTANT ======================
+        // offset for dagRequest should be in accordance with fields
+        for (ColumnRef col: getFields()) {
+          Integer pos = colPosMap.get(col.getColumnInfo());
+          if (pos != null) {
+            TiColumnInfo columnInfo = columnInfoList.get(indexColIds.get(pos));
+            if (col.getColumnInfo().equals(columnInfo)) {
+              dagRequestBuilder.addOutputOffsets(pos);
+              count ++;
+              colOffsetMap.put(col, pos);
+            }
+          }
+          // if a column of field is not contained in index selected,
+          // logically it must be the pk column and
+          // the pkIsHandle must be true. Extra check here.
+          else if (col.getColumnInfo().isPrimaryKey() && tableInfo.isPkHandle()) {
```

Comment:
For index single read case: the fields must be consisted of a subset of index selected and optionally contains primary key if pk is handle.